providers/fips/selftest.c

```diff
@@ -24,8 +35,55 @@
 #define DIGEST_NAME "SHA256"
 
 static int FIPS_state = FIPS_STATE_INIT;
+static CRYPTO_RWLOCK *self_test_lock = NULL;
 static unsigned char fixed_key[32] = { 0 };
 
+static CRYPTO_ONCE fips_self_test_init = CRYPTO_ONCE_STATIC_INIT;
+DEFINE_RUN_ONCE_STATIC(do_fips_self_test_init)
+{
+    self_test_lock = CRYPTO_THREAD_lock_new();
+
+    return self_test_lock != NULL;
+}
+
+/*
+ * We should try to clean up the lock when the fips.so/fips.dll file is
+ * unloaded. On platforms that we know how to detect that we will do so. On
+ * other platforms we will leak the lock. This doesn't really matter though
+ * because the OS should free it anyway.
+ */
+
+#if defined(__GNUC__)
+
+static __attribute__((destructor)) void cleanup(void)
+{
+    CRYPTO_THREAD_lock_free(self_test_lock);
+}
+
+#elif defined(_WIN32) || defined(__CYGWIN__)
+# ifdef __CYGWIN__
+/* pick DLL_[PROCESS|THREAD]_[ATTACH|DETACH] definitions */
+#  include <windows.h>
+/*
+ * this has side-effect of _WIN32 getting defined, which otherwise is
+ * mutually exclusive with __CYGWIN__...
+ */
+# endif
+
+BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved);
```

Comment:
why not use the same mechanism for the ini (i.e DllMain / __attribute((constructor)) ?.

Just as a reminder:
(1) Each time the fips module loads the integrity of the module needs to be checked.
(2) Once during installation the KATS need to also run after the integrity check..

(3) Self tests can be run on demand as a service after the module loads.

Reply:
> why not use the same mechanism for the ini (i.e DllMain / __attribute((constructor)) ?.

Because easier.  *One* codepath for the RUN_ONCE stuff, and no reason to deal with the complexities of diverse ini mechanisms.  The fini stuff is easier to deal with, since we only need to deal with platforms that don't implicitly load modules for the whole process life (implicit `RTLD_NODELETE`)...  this makes me think, though, that for those who do, we should add an atexit() call...  @mattcaswell?

Reply:
I would prefer to see the constructors because they are the mechanisms that are discussed in the FIPS documents.. For  the cases that dont handle this you will know because the state will not be setup when the post gets entered.. (i.e- no constructor has been called)

Reply:
> why not use the same mechanism for the ini (i.e DllMain / __attribute((constructor)) ?.

There isn't a universally portable way of doing this on all platforms. We know how to do it for gcc/clang and Windows. But other platforms...who knows? Doing it this way it works *everywhere*.

We *could* use DllMain/__attribute((constructor)) on platforms where we know it works and fall back to RUN_ONCE on other platforms. But that just adds unnecessary complexity since RUN_ONCE universally works.

> I would prefer to see the constructors because they are the mechanisms that are discussed in the FIPS documents.

I don't think we should tie ourselves up in knots trying to do things exactly the same way as in the FIPS documents. My understanding was that that approach is not mandated.

Reply:
I guess we will see :)

Reply:
Isn't it mandated? We've got this impression from our test lab that the library constructor/destructor approach is actually mandated (if at all supported by the platform).

Reply:
Maybe a question for the lab. I thought this was the case also..

Reply:
@mattcaswell Here is a reference to the docs:
https://csrc.nist.gov/csrc/media/projects/cryptographic-module-validation-program/documents/fips140-2/fips1402ig.pdf

See Section 9.10 Power Up Tests for Software Module Libraries

Reply:
I jut had a read through, and sure enough, it does look mandated...
(which, btw, makes me wonder why the self test code isn't implemented as a constructor / .ini)

Reply:
So the design says we will just 'trigger' it from the constructor.. So it effectively just sets the fips state. We needed to defer so we can pass thru the relevant info to the self test at the right time.

Reply:
Ugh, too early in the morning...  I forget that we do the power on test at installation...

But then, please explain to me why we suddenly care about constructors for *this* PR, with reference to what's mandated in FIPS 140-2 IG?

Reply:
? We care because this PR is trying to do things with run once instead of using constructors.

Reply:
Why?

Reply:
Hmmm. Reading the linked document is does appear to be mandated. But what do we do on platforms where we don't know how to do it this way? Or is FIPS just not supported on such a platform?

Reply:
Its not ideal - but If we know the platform we could fallback to just setting the fips_state in this case.. For now just assume we only support platforms that handle it...

Reply:
> For now just assume we only support platforms that handle it...

Which is a little problematic - because by the time we get to attempting to compile this file, we're already building the FIPS module. We somehow need to know before we start - otherwise we'll end up with an unusable module being built

Reply:
Not  really seeing a issue if you can force it not to build with a config option..

Reply:
I still fail to understand why this i discussed *in this PR*, as if that has any impact on it.

@slontis, you wrote that what we do need to do in a DEP is to set an initial FIPS state (`FIPS_STATE_INIT`, I presume), and defer the rest to the first time `OSSL_provider_init()` is called.  Right?

As far as I can see, this PR is all about that deference, and nothing to do with setting the initial state.

What did I miss?

Reply:
> @slontis, you wrote that what we do need to do in a DEP is to set an initial FIPS state (FIPS_STATE_INIT, I presume), and defer the rest to the first time OSSL_provider_init() is called. Right?

Is that allowed? My understanding from the IG was that it wasn't? If we're allowed to defer the running of the tests to the `OSSL_provider_init()` call, then isn't sufficient just to have the `FIPS_state` variable declared with an initialisation value. Now I'm confused.

Reply:
> I still fail to understand why this i discussed in this PR, as if that has any impact on it.

It very much has an impact on it? The whole RUN_ONCE idea isn't needed if we always run the tests in  DEP.

Reply:
There's a catch: because we want to perform the default KATs *on installation*, i.e. run with the idea that *installation* is the power-on moment, we depend on the integrity test, which among others checks that the "installed" signature exists and is valid.  However, that check can't be performed earlier than on first `OSSL_provider_init()`, because that's how the installation data is passed.

This is in the design.

Reply:
This was discussed in the initial FIPS meeting.
The docs say that the post tests must be 'triggered' in the DEP.
Tim's interpretation of this was that 'trigger' would just mean setting the state.
This is what was decided would be done. The labs did read this design document, so we can assume it is ok.

Reply:
So the conclusion should then be that for *this* PR, the whole DEP thing isn't a concern, only the fini for removing the lock on platforms where this is applicable.

Reply:
The DEP code was also in the design doc, before it was all simplified..

Reply:
If the DEP code has to be there anyway, it may as well go in now and the RUN_ONCE is then not needed.

Reply:
... and for every platform that has a different .ini style, we need to add another piece of code, almost but not quite the same as well the others.

I don't see why it is necessary to add code complexity when the `RUN_ONCE` method works on all

Reply:
> So the conclusion should then be that for this PR, the whole DEP thing isn't a concern,

It seems to me that the design is totally at odds to what I'm reading in the IG.

BTW, if all we need to do is set a variable value to count as the tests being "triggered" then can't that just be done in the variable declaration without a DEP at all? The OS will ensure that that variable is given that value during library load...

Reply:
> I don't see why it is necessary to add code complexity when the RUN_ONCE method works on all

Well that was my starting point for this...but I'm being told above that the RUN_ONCE method isn't allowed by the IG. Confused.

Reply:
Don't assume that FIPS makes sense.
There are hoops - and you have to jump thru them.
Logic isnt going to help you here, thats all I am going to say :)

Reply:
> but I'm being told above that the RUN_ONCE method isn't allowed by the IG.

???????  I didn't read it like that, at all.

Reply:
If it turns out that we can't assume that *installation* equals *power-on*, then there's no reason whatsoever to have the KATs tests depend on the integrity test, as far as I can see.  In that case, we need to redesign the state machine that we have in the design document.

Reply:
> ??????? I didn't read it like that, at all.

They sentences seem to be:

> These nuances are important because while a
> cryptographic module is allowed to intercept calls to a service when the power-up tests are running and
> suppress any output of results until the tests complete, a module is not allowed to initiate the tests from within
> that service. In other words, a software module implemented as a library shall not rely on calls inside any
> function exported as a supported service to initiate the power-up tests.

Reply:
@t-j-h care to comment?

Reply:
So basically, that says that you may not initiate the test via a callback through an external service (such as `RUN_ONCE`). Fair enough. But strictly speaking, this PR doesn't do that, all it does is create a lock, which doesn't initiate the test in and of itself.

Reply:
>  But strictly speaking, this PR doesn't do that, all it does is create a lock, which doesn't initiate the test in and of itself.

No, that's true. But the rest of the function does do that. And this function is called from `OSSL_provider_init`. If that isn't an acceptable solution then it makes this whole PR kind of moot, and a new approach will be needed.

Reply:
The specific mechanism that is used does not have to follow precisely what is stated. 
Alternate implementations are permitted. You just have to work through these with the labs to make sure the underlying requirements are being met.

And remember that there are explicit limitations of what you are meant to safely trigger out of DllMain and friends according to the platform vendors. Nothing substantial is meant to happen in DllMain.

See [https://docs.microsoft.com/en-us/windows/win32/dlls/dllmain](url) and specifically:

> There are significant limits on what you can safely do in a DLL entry point. See General Best Practices for specific Windows APIs that are unsafe to call in DllMain. If you need anything but the simplest initialization then do that in an initialization function for the DLL. You can require applications to call the initialization function after DllMain has run and before they call any other functions in the DLL.


What is stated in the FIPS docs is actually precisely what is not meant to be done. The trigger can be setting a variable. And testing the variable elsewhere. That is allowed. We don't need to overly complicate things here.

Reply:
> The trigger can be setting a variable

@t-j-h Does that have to be done as an explicit entry point? Or is it sufficient just to declare a variable with the "run tests" value, i.e. is it enough to do this:

````
static int FIPS_state = FIPS_STATE_INIT;
````
Or do we have to do this:

````
static int FIPS_state;

static __attribute__((constructor)) void init(void)
{
    FIPS_state = FIPS_STATE_INIT;
}
````

The two are identical functionally, but the latter is problematic because it is non-portable (there isn't a fully portable way of doing this AFAIK)

Reply:
What about something like:

```C
static int FIPS_state = FIPS_STATE_INIT;

#ifdef HAVE_CONSTRUCTOR
static __attribute__((constructor)) void init(void)
{
    FIPS_state = FIPS_STATE_INIT;
}
#endif
```
It's a bit pointless but meets the requirement.

A better alternative might be:

```C
#ifdef HAVE_CONSTRUCTOR
static int FIPS_state;

static __attribute__((constructor)) void init(void)
{
    FIPS_state = FIPS_STATE_INIT;
}
#else
static int FIPS_state = FIPS_STATE_INIT;
#endif
```